src/backend/kvm/mod.rs

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::Loader;
+use crate::backend::kvm::data::{dev_kvm, kvm_version};
+use crate::backend::kvm::mem::Region;
+use anyhow::Result;
+pub use kvm_bindings::kvm_userspace_memory_region as KvmUserspaceMemoryRegion;
+use kvm_ioctls::Kvm;
+use kvm_ioctls::{VcpuFd, VmFd};
+use lset::Span;
+use mmarinus::{perms, Kind, Map};
+use primordial::Page;
+use std::collections::VecDeque;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use x86_64::VirtAddr;
+
+mod builder;
+mod config;
+mod data;
+mod hasher;
+mod mem;
+mod thread;
+
+struct Keep {
+    kvm_fd: Kvm,
+    vm_fd: VmFd,
+    cpu_fds: VecDeque<VcpuFd>,
+    sallyport_blocks: Span<VirtAddr, NonZeroUsize>,
+    regions: Vec<Region>,
+}
+
+impl Keep {
+    pub fn add_memory(&mut self, pages: usize) -> Result<i64> {
+        let mem_size = pages * Page::SIZE;
+        let last_region = self.regions.last().unwrap().as_guest();
+
+        let map = Map::map(mem_size as usize)
+            .anywhere()
+            .anonymously()
+            .known::<perms::ReadWrite>(Kind::Private)?;
+
+        let region_start = map.addr();
+        let region = KvmUserspaceMemoryRegion {
+            slot: self.regions.len() as _,
+            flags: 0,
+            guest_phys_addr: last_region.start.as_u64() + last_region.count,
+            memory_size: mem_size as _,
+            userspace_addr: region_start as _,
+        };
+
+        unsafe {
+            self.vm_fd.set_user_memory_region(region)?;
+        }
+
+        self.regions.push(Region::new(region, map));
+
+        Ok(region_start as _)
+    }
+}
+
+pub struct Backend;
+
+impl crate::backend::Backend for Backend {
+    #[inline]
+    fn name(&self) -> &'static str {
+        "kvm"
+    }
+
+    #[inline]
+    fn shim(&self) -> &'static [u8] {
+        include_bytes!(concat!(env!("OUT_DIR"), "/bin/shim-sev"))
+    }
+
+    #[inline]
+    fn have(&self) -> bool {
+        data::dev_kvm().pass
+    }
+
+    fn data(&self) -> Vec<super::Datum> {
+        vec![dev_kvm(), kvm_version()]
+    }
+
+    #[inline]
+    fn keep(&self, shim: &[u8], exec: &[u8]) -> Result<Arc<dyn super::Keep>> {
+        builder::Builder::load(shim, exec)
+    }
+
+    #[inline]
+    fn hash(&self, shim: &[u8], exec: &[u8]) -> Result<Vec<u8>> {
+        hasher::Hasher::load(shim, exec)
```

Comment:
Just return `Ok(Vec::new())` here and delete the hasher module.

Reply:
done